cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+#include <cassert>
+#include <cstdlib>
+#include <iostream>
+#include <regex>
+#include <tuple>
+#include <string>
+#include <unordered_map>
+
+#include "btx_cudainterval_callbacks.hpp"
+
+#include "context_manager.hpp"
+
+#include <metababel/metababel.h>
+
+
+struct data_s {
+  // store entry state per thread in order to use in exit callback, conditioned
+  // on the cuResult being a non-error
+  std::unordered_map<hpt_t, int64_t> entry_ts;
+  std::unordered_map<hpt_t, std::string> entry_kernel_name;
+  std::unordered_map<hpt_t, size_t> entry_traffic_size;
+
+  // kernels (CUfunction objects) are "created" before use, at which time we save
+  // the name so at execute time we know the name and can emit an event with the name
+  std::unordered_map<hp_kernel_t, std::string> hp_kernel_to_name;
+
+  // At device operation launch time, save the device the operation was launched on,
+  // keyed by the name, since this is not available in the device profiling events
+  std::unordered_map<hpt_function_name_t, CUdevice> hpt_function_name_to_dev;
+
+  // State for device event profiling
+  //
+  // The flow of events for event profiling is:
+  //
+  // 1. device operation entry (cuMemcpy, cuLaunch, etc)
+  // 2. event_profiling
+  // 3. device operation exit
+  // 4. unknown number of intervening events
+  // ...
+  // 5. event profiling results
+  // 
+  // The information needed to emit a thapi device event is spread out throughout
```

Comment:
maybe use the full name of the yaml? Like `lttng::interval_device`? We will do a big sed when we will use correct name for the THAPI babeltrace2 event :)